### ISSUE TITLE
add a CDNs header under use it

### ISSUE THREAD

**Issue by Haroenv** [yarnpkg/website#723]
* unpkg
* jsDelivr

<img width="383" alt="screen shot 2017-11-20 at 10 21 42" src="https://user-images.githubusercontent.com/6270048/33010945-a23cc09a-cddc-11e7-85b3-438fe4909116.png">

**Comment by xtuc** [comment]
Here is also a CDN I maintain: https://bundle.run. It's a bundler a service + a CDN on front of it. Can we add it there?

**Comment by Haroenv** [comment]
sure!

**Comment by Daniel15** [comment]
Don't forget CDNJS, which is probably the largest one :)

**Comment by Haroenv** [comment]
I don’t think CDNjs has a 1:1 npm <-> cdn mapping though, and neither are they putting something in package.json @Daniel15

**Comment by Daniel15** [comment]
@Haroenv There's an Algolia index used on their site (https://cdnjs.com/). Could npm package name be indexed too? Their JSON files have an `npmName` field (see https://github.com/cdnjs/cdnjs/blob/master/documents/autoupdate.md)

**Comment by Haroenv** [comment]
Good point, cdnjs _does_ have an Algolia index, so we could query that with the `npmName` as filter, or simply doing getObject (and dealing with double namings)

**Comment by mjackson** [comment]
I'm a little concerned that the order of CDNs in this list is going to convey a preference to users. For example, on several of the packages that I maintain, like [react-router](https://yarnpkg.com/en/package/react-router), jsDelivr is currently at the top of the list of CDNs. However, we don't actually recommend that people use jsDelivr in our README.

Would it be feasible to do a quick grep through the README to see if it contains any references to a particular CDN before putting it in the list?

**Comment by Haroenv** [comment]
That’s an interesting idea. We have to only make sure that it won’t shuffle the order once it’s already rendered, but the search can happen at indexing time to avoid that.

**Comment by xtuc** [comment]
Just an idea, but what about a new field in the package.json? `CDNs` which are the ONLY CDNs that will show up on the page.

So that @mjackson can only show his unpkg and me only bundle.run :smile:

**Comment by Daniel15** [comment]
> CDNs which are the ONLY CDNs that will show up on the page.

What's the use case though? If a file is available on multiple CDNs, why would some of them intentionally be hidden? I agree that being able to influence the order is good, but I don't understand the purpose of totally hiding particular options.

**Comment by xtuc** [comment]
@Daniel15 ok, you're right.

**Comment by MartinKolarik** [comment]
I think the original idea behind this section was providing links to all existing services so that users could quickly get the link to whichever service they prefer.

In this context, hiding services not mentioned in package.json/readme doesn't make much sense IMHO.

**Comment by Haroenv** [comment]
Yes, adding some order based on appearance in readme seems nice, hiding others would seem not like a good idea

**Comment by xtuc** [comment]
This solution looks good to me. We can order by the number of occurrence in the README.

I just hope nobody has something like "don't use [CDN]" because the CDN will probably show up first in the list.

**Comment by Haroenv** [comment]
Tracking that idea in https://github.com/algolia/npm-search/issues/78

**Comment by mjackson** [comment]
As a package author, I don't personally want a CDN that I haven't tested and recommended to show up on my package at all. The README should contain all the information anyone needs to install and use my package.

**Comment by Daniel15** [comment]
>  I don't personally want a CDN that I haven't tested

Why though? A CDN is just a network of servers that hosts static content. I'm not quite sure what testing would need to be performed? What if a user of the package wants to use CloudFlare/Akamai/CloudFront/MaxCDN/something else? Would that be disallowed in the readme too?

**Comment by Haroenv** [comment]
What if the package author simply doesn't know about the CDN or chose not to include all of the CDNs that work @mjackson? I'm not sure if we can assume exclusive supported CDNs just from the readme, and adding a field to package.json seems weird (what if url changes etc.)

What if there could be an affordance so that if it's mentioned in the readme, it gets first and some icon or similar with info that it was mentioned in the README? I think that would be a good tradeoff

**Comment by MartinKolarik** [comment]
Actually, I personally prefer npm over yarn, so I wonder if we could make the "Use it" section show `npm install` instead of `yarn add` for my packages 😆

**Comment by mjackson** [comment]
Sure, @Haroenv. That sounds like a reasonable trade-off. But I would still expect to find better usage instructions in the README.

For example, the unpkg & jsdelivr links take people to a page that lists all the files in a package, which isn't terribly useful. Which file do you need? Many packages ship multiple versions of their library. For example, the `react` package ships both a [production](https://unpkg.com/react@16.1.1/umd/react.production.min.js) and [development](https://unpkg.com/react@16.1.1/umd/react.development.js) build. But the links under "Use it" don't actually link to a page that tells people how to "use it" at all. 🤡

At least the bundle.run link gives people a file they can use... 😅

**Comment by Haroenv** [comment]
Good point. We can be consistent in leaving off the trailing space

**Comment by MartinKolarik** [comment]
The problem is that many packages have a default file that only works on node.js or don't have a default file at all (this includes popular packages such as react or bootstrap). IMHO it's much better to show all files than a wrong or even broken link. I would also argue that in most cases it's obvious which file is the right one.

**Comment by mjackson** [comment]
React does have a main file, it's just CommonJS. That's why I say that bundle.run is the only CDN that gives people a usable file - because it's the only one that's doing any bundling. 😛

**Comment by Daniel15** [comment]
> React does have a main file, it's just CommonJS. 

Arguably an UMD module is a better main file, as it supports AMD and globals in addition to CommonJS (so any UMD module will also work fine as a CommonJS module). I guess that's likely to change with native ES6 modules though? My knowledge is likely outdated by this point 😛

**Comment by mjackson** [comment]
Honestly, this whole addition just feels to me like an attempt by the jsdelivr people to stay relevant. @MartinKolarik put jsdelivr at the top of the list because he works for them.

How about this: can we at least randomize the order of the list? As I've said before, I think the bundle.run link is probably the best one to have under the "Use it" section because it's the only one that actually provides a usable file. So I honestly wouldn't mind if you removed both jsdelivr and unpkg. But if jsdelivr insists on being included in the list, let's at least make the order random so they don't always look like they're the favorite.

**Comment by xtuc** [comment]
> bundle.run is extremely slow though

Really? The first take you hit a package it needs to run the bundling process (which is super slow) but then it's cached. The next requests are fetched from the cache.

Do you have an example?

**Comment by Haroenv** [comment]
I think @Daniel15 was talking about the first hit to the bundler

**Comment by xtuc** [comment]
Yes it's a known issue, maybe a need to figure out something like a cache warmup thing but that's not the place to discuss about that :)

**Comment by mjackson** [comment]
I'm just saying, if the header says "Use it" the next thing I'd expect to see in the list is probably something like this:

```html
<script src="https://bundle.run/react" crossorigin></script>
```

With a button to click and copy that snippet so you can paste it into your HTML doc. That's probably the most useful thing you can put there.

**Comment by Daniel15** [comment]
@mjackson Right, I get you. Maybe the title "Use it" is misleading and it should be something like "Download the files" or something like that, idk.

@xtuc Yeah, I meant the first hit. The bundle.run website doesn't document the cache eviction policies so I have no idea how often the slow loads will be encountered. That's why I'm hesitant to use it.

**Comment by MartinKolarik** [comment]
@mjackson it's funny to see you trying so hard just to get the order changed. Please don't tell me that if it was you implementing this, you'd have put unpkg last. I'd also like to point out it wasn't *my* idea to add this section.

That said, I'm personally completely fine with randomizing the order if that's what others think is the best solution. We've previously had to do the exact same thing with sponsors on jsDelivr website because they felt the one that comes first looks more important.

**Comment by Daniel15** [comment]
> this whole addition just feels to me like an attempt by the jsdelivr people to stay relevant. 

jsdelivr is still relevant though. It's still very popular, and has features that none of the other services have (eg. APIs for loading usage stats). In my experience, it's usually faster than cdnjs.

**Comment by ryanflorence** [comment]
I don't get why CDNs are even being listed.

**Comment by Haroenv** [comment]
Because multiple people asked for it @ryanflorence

**Comment by ryanflorence** [comment]
What if multiple people ask to remove them? 🤔

It's not really helpful if this website can't know exactly which file on the CDN to use, is it?

**Comment by Haroenv** [comment]
Open a new issue if you want to see them gone. 

I think I’ll close this issue here, the discussion is going on for too long and missing the point. 

This is nothing personally against anyone here, it’s just getting draining on me and I feel attacked for small decisions made

**Comment by Daniel15** [comment]
Maybe we could update the "Browse Files" view to allow copying CDN links for each listed file? Would that solve some of the concerns? So you look at the file list, determine the right file you need (eg. based on the readme), then there's a handy button to copy a link to that exact file.

**Comment by MartinKolarik** [comment]
@Daniel15 We're overthinking this. Many projects have CDN links directly in the readme so if you actually start reading it, you may as well copy the links from there. I know it isn't the case for all projects, but in those cases there are often no install instructions at all. I would also argue that for 90% of projects it is obvious which file you should get.

**Comment by PeterDaveHello** [comment]
CDNJS doesn't have npm to hosted library 1:1 mapping, the `npmName` only works on the libraries using npm registry as auto-update source, though it's not implemented yet, not sure if it's acceptable also to query the repository URL (ignore https, git or ssh protocol prefix) to find the library hosted on CDNJS?

**Comment by Haroenv** [comment]
I guess that would be the only reasonable way to find the cdnjs url. Do you want to contribute @PeterDaveHello ?

**Comment by PeterDaveHello** [comment]
@Haroenv I'm not able to write new codes recently, but please feel free to contribute it here: https://github.com/cdnjs/new-website/blob/master/apiServer.js, I'll review ASAP when I can :)
